A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,643 @@
+    of uncompressed message size.
+
+## Migrate from OpenCensus to OpenTelemetry
+
+### gRPC OpenCensus API
+The existing gRPC OpenCensus tracing APIs in grpc-census plugin are different between
+languages, e.g. in grpc-java it is zero-configuration: as long as the grpc-census
+dependency exists in the classpath, the traces are automatically generated.
+In C++, there is an API for users to call to enable tracing. In Go, it is exposed 
+via stream tracers. The OpenTelemetry API will coexist with the OpenCensus API.
+We keep the grpc-census plugin to allow users who already depend on grpc-census to
+continue using it for newer grpc versions.
+
+gRPC users depending on the grpc-census plugin have non-trivial migration paths
+to OpenTelemetry. Consider the following use cases:
+1. Compatibility between a gRPC client and server as two distributed components,
+where during migration one will use OpenCensus and the other will use OpenTelemetry. 
+2. Migrate an application binary where both OpenCensus and OpenTelemetry maybe exist
+   in the dependency tree. This can be the application’s own tracing code, or gRPC
+   OpenCensus, or other dependencies that involve OpenCensus and/or OpenTelemetry.
+
+Here are the suggested solutions for both use cases. 
+
+### Migrate to OpenTelemetry: Cross-process Networking Concerns
+When users first introduce gRPC OpenTelemetry, for the time window when the
+gRPC client and server have mixed plugins of OpenTelemetry and OpenCensus, 
+spanContext can not directly propagate due to different header name and wire format. 
+To tackle this, gRPC will expose a custom `GrpcTraceBinPropagator` 
+that implements `TextMapPropagator`. This will allow gRPC to keep using `grpc-trace-bin` 
+header for context propagation and also support other propagators.
+When using `grpc-trace-bin` the OpenCensus spanContext and OpenTelemetry spanContext 
+are identical, therefore a gRPC OpenCensus client can speak with a gRPC OpenTelemetry 
+server and vice versa. It is encouraged to use `GrpcTraceBinPropagator` for the migration. 
+Using the same header greatly simplifies rollout. However, there is a caveat about 
+`GrpcTraceBinPropagator`:
+
+Currently, OpenTelemetry propagator API only supports `TextMapPropagator`,
+that is to send string key/value pairs between the client and server, which is
+different from the binary header that gRPC currently uses. The future roadmap
+to support binary propagators at OpenTelemetry is unclear. So, gRPC will use
+propagator API in TextMap format with an optimization path (Go and Java) to work
+around the lack of binary propagator API to support `grpc-trace-bin`. In fact,
+TextMap propagator does not show visible performance impact for C++, which is
+the most sensitive language to performance, based on internal micro benchmarking.
+Therefore, gRPC will only support propagating `grpc-trace-bin` in TextMap propagator.
+Only one `grpc-trace-bin` header will be sent for a single RPC as long as only one of
```

Comment:
@yashykt How are we going to implement this?  Are we just going to have the two CallTracers set the same metadata, and the last one wins?  Or are we going to do something more clever to make sure that we don't need to encode it multiple times?

Reply:
I believe we said that using two plugins that both use the same metadata "grpc-trace-bin" is equivalent to shooting yourself in the foot, and should not be done.